theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,686 @@
+        super(AbstractConv2d, self).__init__(imshp, kshp, bsize,
+                                             border_mode, subsample, filter_flip)
+
+    def make_node(self, img, kern):
+        if img.type.ndim != 4:
+            raise TypeError('img must be 4D tensor')
+        if kern.type.ndim != 4:
+            raise TypeError('kern must be 4D tensor')
+
+        broadcastable=[img.broadcastable[0],
+                       kern.broadcastable[0],
+                       False, False]
+        #output = img.type.__class__(dtype=img.type.dtype,
+        #                            broadcastable=broadcastable)()
+        output = img.type.clone( broadcastable=broadcastable)()
+        return Apply(self, [img, kern], [output])
+
+    def perform(self, node, inp, out_):
+        raise NotImplementedError('AbstractConv2d theano optimization failed')
+
+    def grad(self, inp, grads):
+        bottom, weights = inp
+        top, = grads
+        d_bottom = AbstractConv2d_gradInputs(self.imshp, self.kshp,
+                                             self.bsize,
+                                             self.border_mode,
+                                             self.subsample,
+                                             self.filter_flip)(
+            weights, top, bottom.shape[-2:])
+        d_weights = AbstractConv2d_gradWeights(self.imshp, self.kshp,
+                                               self.bsize,
+                                               self.border_mode,
+                                               self.subsample,
+                                               self.filter_flip)(
+            bottom, top, weights.shape[-2:])
+        return d_bottom, d_weights
+
+
+class AbstractConv2d_gradWeights(BaseAbstractConv2d):
+    """Gradient wrt. filters for `AbstractConv2d`.
+
+    :note: You will not want to use this directly, but rely on
+           Theano's automatic differentiation or graph optimization to
+           use it as needed.
+
+    """
+
+    def __init__(self,
+                 imshp=None,
+                 kshp=None,
+                 bsize=None,
+                 border_mode="valid",
+                 subsample=(1, 1),
+                 filter_flip=True):
+        super(AbstractConv2d_gradWeights, self).__init__(imshp, kshp, bsize,
+                                                         border_mode, subsample, filter_flip)
+
+    ## Update shape/height_width
+    def make_node(self, img, topgrad, shape):
+        if img.type.ndim != 4:
+            raise TypeError('img must be 4D tensor')
+        if topgrad.type.ndim != 4:
+            raise TypeError('topgrad must be 4D tensor')
+        if self.subsample != (1, 1) or self.border_mode == "half":
+            if shape is None:
+                raise ValueError('shape must be given if subsample != (1, 1)'
+                                 ' or border_mode == "half"')
+
+        shape = as_tensor_variable(shape)
```

Comment:
I guess it was copied from `GpuCorrMM_gradWeights`, where the shape is not passed to `Apply` if we're able to compute it in the apply code (i.e., when we're not subsampling and border_mode is not "half").
It seems that support for `shape=None` was removed now, so the `if` statements and `ValueError` are dead code that should be removed.

> We could also use `self.kshp` if it is specified.

That's still possible, would that help performance somewhere? I.e., using `as_tensor_variable(self.kshp)` if that is fully specified? (Although it's unclear to me if `kshp` includes all 4 dimensions or is the crippled version from the legacy Op.)